### ISSUE TITLE
.Renv / requirements.txt parity

### ISSUE THREAD

**Issue by isabelizimm** [tidymodels/vetiver-python#7]
In both vetiver-r and vetiver-python, the metadata of a vetiver model holds a minimal set of dependencies needed to make a prediction.

For R, the `.Renv` generates a lockfile of all dependencies, and then only the packages needed to make a prediction are saved in the metadata.

In Python, a requirements.txt must be generated. The big options here might be `pip freeze`, `pip-compile`, or generating this file manually from a pre-set list of requirements. The benefits of the `pip` options is that there would be a defined version number, to save users headache in case different metadata is used in new vetiver versions. This seems preferable to a list with unspecified version numbers. There might need to be some parsing to either `pip freeze` or `pip-compile` to specify only the packages needed for a prediction, rather than all packages a user has installed. 

It seems that the "best parity" option might to be to generate a large file via pip-compile, then grep (or some other parsing) to only save the requirements specified from a set list (ie, vetiver, pins, and torch/sklearn/etc). (WIP solution)

Is there any other useful context I am missing?

**Comment by juliasilge** [comment]
I would just clarify that the order is the other way in the R side: first, the packages needed to make a prediction are saved in the `vetiver_model()` metadata, and then we use renv to create a scoped-down `renv.lock` that only tracks the packages + versions needed for prediction.

**Comment by isabelizimm** [comment]
> I would just clarify that the order is the other way in the R side: first, the packages needed to make a prediction are saved in the `vetiver_model()` metadata, and then we use renv to create a scoped-down `renv.lock` that only tracks the packages + versions needed for prediction.

Thank you for the clarification! I updated the text above to match your comment and avoid further confusion :)

**Comment by isabelizimm** [comment]
Looking into `pip-compile` some more since it seems to make the most sense for this use case. We could create a temporary file from the `VetiverModel` metadata that contains the package names with no versions, then `pip-compile tempfile --output-file=requirements.txt` to generate a `requirements.txt` that is versioned. 

I tried this out, and the generated output looks something like below. You get an ugly tempfile name (fixable, but also might want to keep it in this format so users don't try to look for a file that has been deleted?), but I think the rest of the functionality is similar to what we want. Other features include support for installing from GitHub and support for [hash-checking mode](https://pip-tools.readthedocs.io/en/latest/#using-hashes), if wanted.

`#
# This file is autogenerated by pip-compile with python 3.8
# To update, run:
#
#    pip-compile --output-file=requirements.txt '<tempfile._TemporaryFileWrapper object at 0x107c00310>'
#
joblib==1.1.0
    # via scikit-learn
numpy==1.22.3
    # via
    #   pandas
    #   scikit-learn
pandas==1.4.2
    # via -r <tempfile._TemporaryFileWrapper object at 0x107c00310>
...` truncated for brevity. 

CC: @machow

**Comment by machow** [comment]
I wonder if having concrete examples could help here? What do you think about us setting up a repo with these things?:

* a folder (`vetiver-r-<something>`) example script + output of R vetiver's dockerfile or plumber.R generated files
* a folder (`vetiver-py-<something>`) example script + output of py version of above (based on current draft etc..)
* a folder (`pins-board`) with the data they are using

This way, a person can clone that repo and do something like below, and commit the results...

**python**

```python
from pins import board_folder

board = board_folder("pins-board")
board.pin_read(...)
board.pin_write(...)
```

**R**

```R
library(pins)

board <- board_folder("pins-board")
```

(Alternatively, could see having each repo put these somewhere?)

**Comment by isabelizimm** [comment]
> So something similar for the generated Dockerfile / related files might help here?

Agreed, I do think examples will help clear things up. In #8 I added a generated [Dockerfile](https://github.com/isabelizimm/vetiver-python/blob/write-docker-file/examples/coffeeratings/Dockerfile), [requirements.txt](https://github.com/isabelizimm/vetiver-python/blob/write-docker-file/examples/coffeeratings/vetiver_req.txt), and [app.py](https://github.com/isabelizimm/vetiver-python/blob/write-docker-file/examples/coffeeratings/app.py), the parallel file to plumber.R. I've seen other python packages have an `examples/` directory, so thought this might be a good way to keep these files easily discoverable.

The main difference in the Python/R experience here is the existence of the `requirements.txt` file. It could technically be abstracted away for Docker, but it's needed for RSConnect, so it's probably useful to have the opinionated workflow the same.

**Comment by juliasilge** [comment]
For the R side, as of right now, you can look at this folder to see generated [Dockerfile, renv.lock, and plumber.R files](https://github.com/juliasilge/vetiverdemo/tree/main/netflix-descriptions). In the R package, there are examples of these in tests as well.
